File: bridge/python/src/destack/_impl/mir/metadata/frame.py
from __future__ import annotations

from destack._impl.model import ModelImpl


class FrameTableImpl(ModelImpl):
    """Methods for frame tables."""

    def layout(self, layout):
        """Return one frame layout by id."""

        return sequence_get(self.layouts, layout)

    def materialization(self, state):
        """Return one frame materialization by state id."""

        return sequence_get(self.materializations, state)


class FrameMaterializationImpl(ModelImpl):
    """Methods for frame materialization plans."""

    def slots(self):
        """Return each source frame slot once."""

        return self.copied_slots


class FrameLayoutImpl(ModelImpl):
    """Methods for frame layouts."""

    def slot(self, id_):
        """Return one frame slot by id."""

        return sequence_get(self.slots, id_)

    def value_slot_id(self, value: int):
        """Return the frame slot id for one SSA value."""

        if value >= self.value_count:
            return None

        return value

    def local_slot_id(self, local: int):
        """Return the frame slot id for one local."""

        if local >= self.local_count:
            return None

        return self.value_count + local

    def value(self, value: int):
        """Return the value slot at one SSA value index."""

        if value >= self.value_count:
            return None

        return sequence_get(self.slots, value)

    def local(self, local: int):
        """Return the local slot at one local index."""

        if local >= self.local_count:
            return None

        index = self.value_count + local

        return sequence_get(self.slots, index)

    def value_for_slot(self, id_) -> int | None:
        """Return the SSA value addressed by one slot id."""

        if id_ < self.value_count:
            return id_

        return None

    def local_for_slot(self, id_) -> int | None:
        """Return the local addressed by one slot id."""

        if id_ < self.value_count:
            return None

        local = id_ - self.value_count

        return local if local < self.local_count else None

    def is_environment_slot(self, id_) -> bool:
        """Return whether one slot id addresses the callable environment."""

        return self.environment_slot is not None and self.environment_slot == id_

    def environment(self):
        """Return the callable environment slot when present."""

        if self.environment_slot is None:
            return None

        return self.slot(self.environment_slot)

    def values(self):
        """Return all value slots."""

        return self.slots[: self.value_count]

    def locals(self):
        """Return all local slots."""

        start = self.value_count
        end = start + self.local_count

        return self.slots[start:end]

    def slot_len(self) -> int:
        """Return the frame slot count."""

        return len(self.slots)

    def slot_ids(self):
        """Return all frame slot ids."""

        from destack._generated.mir.metadata.frame import FrameSlotId

        return tuple(range(len(self.slots)))


def sequence_get(sequence, index: int):
    """Return one sequence item when the index is present."""

    if 0 <= index < len(sequence):
        return sequence[index]

    return None

File: client/typescript/src/_impl/mir/metadata/frame.ts
import type {
    FrameLayout,
    FrameLayoutId,
    FrameMaterialization,
    FrameSlot,
    FrameSlotId,
    FrameStateId,
    FrameTable,
} from "../../../_generated/mir/metadata/frame.js";

export const FrameTableImpl = {
    /** Return one frame layout by id. */
    layout(table: FrameTable, layout: FrameLayoutId): FrameLayout | undefined {
        return table.layouts[layout];
    },

    /** Return one frame materialization by state id. */
    materialization(table: FrameTable, state: FrameStateId): FrameMaterialization | undefined {
        return table.materializations[state];
    },
};

export const FrameMaterializationImpl = {
    /** Return each source frame slot once. */
    slots(materialization: FrameMaterialization): ReadonlyArray<FrameSlotId> {
        return materialization.copiedSlots;
    },
};

export const FrameLayoutIdImpl = {
    /** Return the raw frame layout id. */
    raw(id: FrameLayoutId): number {
        return id;
    },
};

export const FrameSlotIdImpl = {
    /** Return the raw frame slot id. */
    raw(id: FrameSlotId): number {
        return id;
    },
};

export const FrameStateIdImpl = {
    /** Return the raw frame state id. */
    raw(id: FrameStateId): number {
        return id;
    },
};

export const FrameLayoutImpl = {
    /** Return one frame slot by id. */
    slot(layout: FrameLayout, id: FrameSlotId): FrameSlot | undefined {
        return layout.slots[id];
    },

    /** Return the frame slot id for one SSA value. */
    valueSlotId(layout: FrameLayout, value: number): FrameSlotId | undefined {
        return value < layout.valueCount ? value : undefined;
    },

    /** Return the frame slot id for one local. */
    localSlotId(layout: FrameLayout, local: number): FrameSlotId | undefined {
        return local < layout.localCount ? layout.valueCount + local : undefined;
    },

    /** Return the value slot at one SSA value index. */
    value(layout: FrameLayout, value: number): FrameSlot | undefined {
        return value < layout.valueCount ? layout.slots[value] : undefined;
    },

    /** Return the local slot at one local index. */
    local(layout: FrameLayout, local: number): FrameSlot | undefined {
        if (local >= layout.localCount) {
            return undefined;
        }

        const index = layout.valueCount + local;

        return layout.slots[index];
    },

    /** Return the SSA value addressed by one slot id. */
    valueForSlot(layout: FrameLayout, id: FrameSlotId): number | undefined {
        return id < layout.valueCount ? id : undefined;
    },

    /** Return the local addressed by one slot id. */
    localForSlot(layout: FrameLayout, id: FrameSlotId): number | undefined {
        if (id < layout.valueCount) {
            return undefined;
        }

        const local = id - layout.valueCount;

        return local < layout.localCount ? local : undefined;
    },

    /** Return whether one slot id addresses the callable environment. */
    isEnvironmentSlot(layout: FrameLayout, id: FrameSlotId): boolean {
        return layout.environmentSlot === id;
    },

    /** Return the callable environment slot when present. */
    environment(layout: FrameLayout): FrameSlot | undefined {
        return layout.environmentSlot === undefined
            ? undefined
            : FrameLayoutImpl.slot(layout, layout.environmentSlot);
    },

    /** Return all value slots. */
    values(layout: FrameLayout): ReadonlyArray<FrameSlot> {
        return layout.slots.slice(0, layout.valueCount);
    },

    /** Return all local slots. */
    locals(layout: FrameLayout): ReadonlyArray<FrameSlot> {
        const start = layout.valueCount;
        const end = start + layout.localCount;

        return layout.slots.slice(start, end);
    },

    /** Return the frame slot count. */
    slotLen(layout: FrameLayout): number {
        return layout.slots.length;
    },

    /** Return all frame slot ids. */
    slotIds(layout: FrameLayout): ReadonlyArray<FrameSlotId> {
        return layout.slots.map((_slot, value) => value);
    },
};
